read-fonts/src/tables/postscript/charstring.rs

```diff
@@ -1,203 +1,644 @@
+/// item variation store, then `blend_state` must be provided.
+pub fn evaluate(
+    charstring_data: &[u8],
+    global_subrs: Index,
+    subrs: Option<Index>,
+    blend_state: Option<BlendState>,
+    sink: &mut impl CommandSink,
+) -> Result<(), Error> {
+    let mut evaluator = Evaluator::new(global_subrs, subrs, blend_state);
+    evaluator.evaluate(charstring_data, Fixed::ZERO, Fixed::ZERO, sink, 0)?;
+    Ok(())
 }
 
-impl From<Operator> for Token {
-    fn from(value: Operator) -> Self {
-        Self::Operator(value)
-    }
+struct Evaluator<'a> {
+    global_subrs: Index<'a>,
+    subrs: Option<Index<'a>>,
+    blend_state: Option<BlendState<'a>>,
+    is_open: bool,
+    have_read_width: bool,
+    stem_count: usize,
+    stack: Stack,
 }
 
-impl<T> From<T> for Token
-where
-    T: Into<Number>,
-{
-    fn from(value: T) -> Self {
-        Self::Operand(value.into())
+impl<'a> Evaluator<'a> {
+    fn new(
+        global_subrs: Index<'a>,
+        subrs: Option<Index<'a>>,
+        blend_state: Option<BlendState<'a>>,
+    ) -> Self {
+        Self {
+            global_subrs,
+            subrs,
+            blend_state,
+            is_open: false,
+            have_read_width: false,
+            stem_count: 0,
+            stack: Stack::new(),
+        }
     }
-}
 
-/// Given a byte slice containing charstring data, returns an iterator yielding
-/// raw operands and operators.
-///
-/// This does not perform any evaluation on the charstring.
-pub fn tokens(charstring_data: &[u8]) -> impl Iterator<Item = Result<Token, Error>> + '_ + Clone {
-    let mut cursor = crate::FontData::new(charstring_data).cursor();
-    std::iter::from_fn(move || {
-        if cursor.remaining_bytes() == 0 {
-            None
-        } else {
-            Some(parse_token(&mut cursor))
+    fn evaluate(
+        &mut self,
+        charstring_data: &[u8],
+        mut x: Fixed,
+        mut y: Fixed,
+        sink: &mut impl CommandSink,
+        nesting_depth: u32,
+    ) -> Result<(Fixed, Fixed), Error> {
+        if nesting_depth > NESTING_DEPTH_LIMIT {
+            return Err(Error::CharstringNestingDepthLimitExceeded);
         }
-    })
+        use ops::*;
+        let mut cursor = crate::FontData::new(charstring_data).cursor();
+        while cursor.remaining_bytes() != 0 {
+            let op = cursor.read::<u8>()?;
+            if op == ESCAPE {
+                let two_byte_op = cursor.read::<u8>()?;
+                // The following "flex" operators are intended to emit
+                // either two curves or a straight line depending on
+                // a "flex depth" parameter and the distance from the
+                // joining point to the chord connecting the two
+                // end points. In practice, we just emit the two curves,
+                // following FreeType:
+                // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L335>
```

Comment:
Optional, wdyt about adding something like

```rust
    pub fn get_multiple_fixed<const COUNT: usize>(&self, first_index: usize) -> Result<[Fixed; COUNT], Error> {
        let mut result = [Fixed::from_i32(0); COUNT];
        for i in 0..COUNT {
            result[i] = self.get_fixed(first_index + i)?;
        }
        Ok(result)
    }
```

So you can write things like

```rust
                        let stack: [Fixed; 6] = self.stack.get_multiple_fixed(0)?;
                        let dx1 = x + stack[0];
                        let dy1 = y;
                        let dx2 = dx1 + stack[1];
```

Compiler will check you don't walk off the end, little bit less noise in the endless streams of dxn = something + stack.

Reply:
Yep, ty! I like this and it's very similar to what I did for handling variation deltas in the COLR table. Done.